GH-23221: [C++] Add support for building with Emscripten

joemarshall opened the pull request:
Split from #37696 

This is just the cmake changes to enable building on emscripten.

Changes are:
1) Support for target system "emscripten"
2) Cmake preset for building libarrow ` ninja-release-python-emscripten` (same as `ninja-release-python`, but with emscripten support)
3) Override file for cmake on emscripten, to set various build parameters that need setting to make it build there.
4) Changes in pyarrow cmake so it works if you are building libarrow as shared library, and also an option to enable the cmake file there to just dump the current arrow configuration, which is useful for cross-compile builds.

* Closes: #23221

joemarshall commented:
@kou here are the cmake changes for building libarrow in emscripten.

kou commented on cpp/CMakePresets.json:
```diff
@@ -46,6 +46,32 @@
       }
     },
+    {
+      "name": "emscripten-overrides",
```
How about rename this to `base-emscripten` or `features-emscripten`?

kou commented on cpp/CMakePresets.json:
```diff
@@ -46,6 +46,32 @@
+        "ARROW_JEMALLOC": "OFF",
+        "ARROW_S3": "OFF",
+        "ARROW_DEPENDENCY_SOURCE": "BUNDLED",
+        "ZLIB_SOURCE": "SYSTEM",
```
Could you sort this list in alphabetical order?

kou commented on cpp/CMakePresets.json:
```diff
@@ -46,6 +46,32 @@
+        "ARROW_SUBSTRAIT": "OFF",
+        "ARROW_DEPENDENCY_USE_SHARED": "OFF",
+        "ARROW_WITH_BROTLI": "OFF",
+        "ARROW_SIMD_LEVEL":"NONE",
```
```suggestion
        "ARROW_SIMD_LEVEL": "NONE",
```

kou commented on cpp/CMakePresets.json:
```diff
@@ -46,6 +46,32 @@
+        "ARROW_DEPENDENCY_USE_SHARED": "OFF",
+        "ARROW_WITH_BROTLI": "OFF",
+        "ARROW_SIMD_LEVEL":"NONE",
+        "ARROW_RUNTIME_SIMD_LEVEL":"NONE",
```
```suggestion
        "ARROW_RUNTIME_SIMD_LEVEL": "NONE",
```

kou commented on cpp/CMakePresets.json:
```diff
@@ -46,6 +46,32 @@
+        "ARROW_WITH_BROTLI": "OFF",
+        "ARROW_SIMD_LEVEL":"NONE",
+        "ARROW_RUNTIME_SIMD_LEVEL":"NONE",
+        "CMAKE_C_BYTE_ORDER":"LITTLE_ENDIAN",
```
```suggestion
        "CMAKE_C_BYTE_ORDER": "LITTLE_ENDIAN",
```

kou commented on cpp/CMakePresets.json:
```diff
@@ -46,6 +46,32 @@
+        "ARROW_SIMD_LEVEL":"NONE",
+        "ARROW_RUNTIME_SIMD_LEVEL":"NONE",
+        "CMAKE_C_BYTE_ORDER":"LITTLE_ENDIAN",
+        "ARROW_WITH_OPENTELEMETRY":"OFF",
```
```suggestion
        "ARROW_WITH_OPENTELEMETRY": "OFF",
```

kou commented on cpp/CMakePresets.json:
```diff
@@ -395,6 +421,16 @@
       "cacheVariables": {}
     },
+    {
+      "name": "ninja-release-emscripten-python",
```
How about adding `ninja-debug-emscripten` and `ninja-release-emscripten` too?

kou replied:
How about renaming this to `ninja-release-python-emscripten` because we have `ninja-release-python-maximal`?

kou commented on cpp/CMakePresets.json:
```diff
@@ -395,6 +421,16 @@
+        "base-release",
+        "features-python"
+      ],
+      "displayName": "Release build which builds an emscripten library, plus PyArrow for Pyodide",
```
Ah, we don't include any PyArrow related features in Arrow C++ now. (GH-14117)
`features-python` is just for enabling Arrow C++ features that are used from PyArrow.

kou reviewed (commented):
(Sorry. My review isn't completed yet. I'll continue later.)

kou commented on cpp/cmake_modules/Emscripten/Platform/EmscriptenOverrides.cmake:
```diff
@@ -0,0 +1,56 @@
+get_filename_component(PLATFORM_FOLDER_PARENT ${CMAKE_CURRENT_LIST_DIR} DIRECTORY)
+list(APPEND CMAKE_MODULE_PATH "${PLATFORM_FOLDER_PARENT}")
+
+include($ENV{EMSDK}/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake)
```
`Emscripten.cmake` exists at `/usr/share/emscripten/cmake/Modules/Platform/Emscripten.cmake` on Debian. ( https://packages.debian.org/bookworm/all/emscripten/filelist )
So this doesn't work on Debian.

It seems that Emscripten provides `emcmake`. Can we use it like `emcmake cmake ...` instead of specifying our toolchain file by `CMAKE_TOOLCHAIN_FILE` in `CMakePresets.json`?

joemarshall replied:
Okay, have done

kou commented on cpp/cmake_modules/Emscripten/Platform/EmscriptenOverrides.cmake:
```diff
@@ -0,0 +1,56 @@
+
+# ensure zlib is built with -fpic
+# and force us to link to the version in emscripten ports
+if(NOT EXISTS ${EMSCRIPTEN_SYSROOT}/lib/wasm32-emscripten/pic/libz.a)
+  execute_process(COMMAND embuilder --pic --force build zlib)
+endif()
+set(ZLIB_LIBRARY ${EMSCRIPTEN_SYSROOT}/lib/wasm32-emscripten/pic/libz.a)
```
Can we do this in `build_zlib()` in `ThirdpartyToolchain.cmake`?

joemarshall replied:
have done

kou commented on cpp/cmake_modules/Emscripten/Platform/EmscriptenOverrides.cmake:
```diff
@@ -0,0 +1,56 @@
+set(ZLIB_LIBRARY ${EMSCRIPTEN_SYSROOT}/lib/wasm32-emscripten/pic/libz.a)
+
+# # override default in emscripten which is to not use shared libs
+set_property(GLOBAL PROPERTY TARGET_SUPPORTS_SHARED_LIBS TRUE)
```
Why do we need this? What shared libraries are used?

joemarshall replied:
python modules are built as shared libraries

kou commented on cpp/cmake_modules/Emscripten/Platform/EmscriptenOverrides.cmake:
```diff
@@ -0,0 +1,56 @@
+# is run, which does bad things like disabling shared libraries
+set(CMAKE_SYSTEM_NAME EmscriptenOverrides)
+
+set(CMAKE_C_FLAGS "-sUSE_ZLIB=1 -sSIDE_MODULE=1 -fPIC -fexceptions")
+set(CMAKE_CXX_FLAGS "-sUSE_ZLIB=1 -sSIDE_MODULE=1 -fPIC -fexceptions")
```
Why do we need `-sUSE_ZLIB=1`?

Could you tell me the document URL for `SIDE_MODULE=1`?

Why do we need to specify `-fPIC` explicitly? CMake's `POSITION_INDEPENDENT_CODE` https://cmake.org/cmake/help/latest/prop_tgt/POSITION_INDEPENDENT_CODE.html doesn't work?

Do we need `-fexception` for `C_FLAGS`?

Is `-fwasm-exceptions` https://emscripten.org/docs/porting/exceptions.html#webassembly-exception-handling-based-support better?

Can we set this in `SetupCxxFlags.cmake` instead of here because `CMAKE_CXX_FLAGS` may be overwritten by users.

joemarshall replied:
have moved this stuff to SetupCXXFlags.

`-sUSE_ZLIB` means to use it from emscripten ports. This is recommended on emscripten.
https://emscripten.org/docs/compiling/Building-Projects.html

joemarshall replied:
I'm not sure about POSITION_INDEPENDENT_CODE, but emscripten is weird, so I specified it explicitly here (and it would be another thing to pass through in thirdpartytoolchain.)

kou commented on cpp/cmake_modules/Emscripten/Platform/EmscriptenOverrides.cmake:
```diff
@@ -0,0 +1,56 @@
+set(CMAKE_C_FLAGS "-sUSE_ZLIB=1 -sSIDE_MODULE=1 -fPIC -fexceptions")
+set(CMAKE_CXX_FLAGS "-sUSE_ZLIB=1 -sSIDE_MODULE=1 -fPIC -fexceptions")
+
+#set(PYARROW_CPP_HOME "$ENV{ARROW_HOME}/lib")
+#list(APPEND CMAKE_FIND_ROOT_PATH "${CMAKE_INSTALL_PREFIX}/cmake")
```
Could you remove them?

kou commented on cpp/cmake_modules/Emscripten/Platform/EmscriptenOverrides.cmake:
```diff
@@ -0,0 +1,56 @@
+#set(PYARROW_CPP_HOME "$ENV{ARROW_HOME}/lib")
+#list(APPEND CMAKE_FIND_ROOT_PATH "${CMAKE_INSTALL_PREFIX}/cmake")
+
+set(Python3_INCLUDE_DIR $ENV{PYTHONINCLUDE})
+set(Python3_LIBRARY $ENV{CPYTHONLIB})
+set(Python3_NumPy_INCLUDE_DIR $ENV{NUMPY_LIB}/core/include)
+set(Python3_EXECUTABLE)
```
Why do we need set them?

joemarshall replied:
This is for pyarrow builds because cmake doesn't always reliably find the correct (cross-compiling) python files. In theory it should work if you just setup the cross compile environment variables correctly and then tell cmake to find python, but in practice, when you have a cross compile version of python, alongside a host version, things are way less likely to fail if one gives it all these hints (and makes sure it knows that we don't have a python executable that cmake can run).

kou replied:
OK. Then how about set them in PyArrow side?
Apache Arrow C++ doesn't use Python. (Apache Arrow C++ just uses Python to run some utility scripts such as linter.)

kou commented on cpp/cmake_modules/Emscripten/Platform/EmscriptenOverrides.cmake:
```diff
@@ -0,0 +1,56 @@
+set(Python3_LIBRARY $ENV{CPYTHONLIB})
+set(Python3_NumPy_INCLUDE_DIR $ENV{NUMPY_LIB}/core/include)
+set(Python3_EXECUTABLE)
+set(CMAKE_SHARED_LIBRARY_CREATE_C_FLAGS "-sUSE_ZLIB=1 -sWASM_BIGINT=1 -fexceptions")
+set(CMAKE_SHARED_LIBRARY_CREATE_CXX_FLAGS "-sUSE_ZLIB=1 -sWASM_BIGINT=1 -fexceptions")
+set(CMAKE_SHARED_LINKER_FLAGS "-sUSE_ZLIB=1 -sWASM_BIGINT=1 -fexceptions")
```
Do we need them?

joemarshall replied:
I slimmed down slightly on the move to setupcxxflags but yes definitely need this stuff to permeate everywhere in the build, so this is the most sensible way to do it.

kou commented on cpp/cmake_modules/Emscripten/Platform/EmscriptenOverrides.cmake:
```diff
@@ -0,0 +1,56 @@
+set(CMAKE_SHARED_LIBRARY_CREATE_C_FLAGS "-sUSE_ZLIB=1 -sWASM_BIGINT=1 -fexceptions")
+set(CMAKE_SHARED_LIBRARY_CREATE_CXX_FLAGS "-sUSE_ZLIB=1 -sWASM_BIGINT=1 -fexceptions")
+set(CMAKE_SHARED_LINKER_FLAGS "-sUSE_ZLIB=1 -sWASM_BIGINT=1 -fexceptions")
+set(CMAKE_STRIP FALSE)
```
Why do we need this?

joemarshall replied:
stripping doesn't work on emscripten - not sure why

kou commented on cpp/cmake_modules/Emscripten/Platform/EmscriptenOverrides.cmake:
```diff
@@ -0,0 +1,56 @@
+set(CMAKE_SHARED_LINKER_FLAGS "-sUSE_ZLIB=1 -sWASM_BIGINT=1 -fexceptions")
+set(CMAKE_STRIP FALSE)
+
+set(ENV{_PYTHON_SYSCONFIGDATA_NAME} $ENV{SYSCONFIG_NAME})
```
Can we remove this?

joemarshall replied:
gone now

kou commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -24,7 +24,9 @@ include(CheckCXXSourceCompiles)
 
 if(NOT DEFINED ARROW_CPU_FLAG)
-  if(CMAKE_SYSTEM_PROCESSOR MATCHES "AMD64|amd64|X86|x86|i[3456]86|x64")
+  if(CMAKE_SYSTEM_NAME MATCHES "Emscripten")
```
Can we use `STREQUAL` instead of `MATCHES`?

kou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1614,6 +1619,9 @@ macro(build_thrift)
     list(APPEND THRIFT_CMAKE_ARGS "-DBOOST_ROOT=${BOOST_ROOT}")
   endif()
+  if(DEFINED Boost_INCLUDE_DIR)
+    list(APPEND THRIFT_CMAKE_ARGS "-DBoost_INCLUDE_DIR=${Boost_INCLUDE_DIR}")
```
How about using `BOOST_INCLUDE_DIR` instead of `Boost_INCLUDE_DIR` because `BOOST_...` is a hint variable https://cmake.org/cmake/help/latest/module/FindBoost.html#hints and `Boost_...` is a cache result variable https://cmake.org/cmake/help/latest/module/FindBoost.html#cache-variables ?

```suggestion
    list(APPEND THRIFT_CMAKE_ARGS "-DBOOST_INCLUDE_DIR=${Boost_INCLUDE_DIR}")
```

joemarshall replied:
I passed the result var as it short circuits a load of logic in findboost which was breaking. I suspect that's why the namespace is passed similarly.

kou commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+.. highlight:: console .. _developers-cpp-emscripten:
+
+#################################################
+ Cross compiling for Webassembly with Emscripten
```
```suggestion
Cross compiling for WebAssembly with Emscripten
```

kou commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+***************
+ Prerequisites
+***************
+You need cmake and compilers etc. installed as per the normal build instructions. Before building with emscripten, you also need to install emscripten and
```
```suggestion
You need CMake and compilers etc. installed as per the normal build instructions. Before building with Emscripten, you also need to install Emscripten and
```

kou commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+   ./emsdk activate <version>
+   source ./emsdk_env.sh
+
+If you want to build pyarrow for `pyodide <https://pyodide.org>`_, you
```
```suggestion
If you want to build PyArrow for `Pyodide <https://pyodide.org>`_, you
```

kou commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+
+If you want to build pyarrow for `pyodide <https://pyodide.org>`_, you
+need ``pyodide-build`` installed via ``pip``, and to be running with the
+same version of python that pyodide is built for, along with the same
```
```suggestion
same version of Python that Pyodide is built for, along with the same
```

kou commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+   cmake --preset "ninja-release-emscripten-python"
+   ninja install
+
+This will install a built static library version of libarrow it into the
```
```suggestion
This will install a built static library version of ``libarrow`` it into the
```

kou commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+   ninja install
+
+This will install a built static library version of libarrow it into the
+emscripten sysroot cache, meaning you can build things that depend on it
```
```suggestion
Emscripten sysroot cache, meaning you can build things that depend on it
```

kou commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+
+This will install a built static library version of libarrow it into the
+emscripten sysroot cache, meaning you can build things that depend on it
+and they will find libarrow.
```
```suggestion
and they will find ``libarrow``.
```

kou commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+emscripten sysroot cache, meaning you can build things that depend on it
+and they will find libarrow.
+
+e.g. if you want to build for pyodide, run the commands above, and then
```
```suggestion
e.g. if you want to build for Pyodide, run the commands above, and then
```

kou commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+ Manual Build
+**************
+
+If you want to manually build for emscripten, take a look at the
```
```suggestion
If you want to manually build for Emscripten, take a look at the
```

kou commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+**************
+
+If you want to manually build for emscripten, take a look at the
+CMakePresets.json file in the arrow/cpp directory for a list of things
```
```suggestion
``CMakePresets.json`` file in the ``arrow/cpp`` directory for a list of things
```

kou commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+   ``arrow/cpp/cmake_modules/Emscripten/Platform/EmscriptenOverrides.cmake``
+
+#. You will quite likely need to set ``ARROW_ENABLE_THREADING`` to ``OFF``
+   for builds targeting single threaded emscripten environments such as
```
```suggestion
   for builds targeting single threaded Emscripten environments such as
```

kou reviewed (commented):
(My first review is completed.)

jorisvandenbossche commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+.. under the License.
+
+
+.. highlight:: console .. _developers-cpp-emscripten:
+
```
```suggestion
.. highlight:: console

.. _developers-cpp-emscripten:
```

jorisvandenbossche commented on cpp/CMakePresets.json:
```diff
@@ -46,6 +46,32 @@
+        "ARROW_S3": "OFF",
+        "ARROW_DEPENDENCY_SOURCE": "BUNDLED",
+        "ZLIB_SOURCE": "SYSTEM",
+        "ARROW_IPC": "OFF",
```
IPC is not only used for network functionality (I see you mentioned below that "``ARROW_IPC`` and anything else that uses network probably won't work."), so this can be enabled by default? Then, for example, reading an Arrow IPC file should work.

joemarshall replied:
oh yeah, my mistake, I was thinking of flight.

joemarshall commented:
Right - updated now - it now minimises what it does in the override file, with the important stuff being split out into setupcxxflags etc.

It also now builds IPC, as suggested by @jorisvandenbossche above, and also acero and substrait, and I fixed building of Brotli.

To make that stuff work, I had to build a host build of protoc in addition to the build for the target platform - this will mean that if you cross compile for other platforms it should now work also.

Python preset is now named ninja-release-python-emscripten as suggested by @hou 

Oh and I updated the PR description above.

kou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1365,16 +1398,51 @@ macro(build_brotli)
+                                         ${BROTLI_BUILD_BYPRODUCTS}
+                        CMAKE_ARGS ${BROTLI_CMAKE_ARGS}
+                        STEP_TARGETS headers_copy)
+  endif()
```
I'm trying this with `--preset=ninja-debug-emscripten`. I need at least the following on Debian GNU/Linux sid:

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index b50f3f1f1..9cd6e6727 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -936,11 +936,15 @@ set(EP_COMMON_TOOLCHAIN "-DCMAKE_C_COMPILER=${EP_C_COMPILER}"
                         "-DCMAKE_CXX_COMPILER=${EP_CXX_COMPILER}")
 
 if(CMAKE_AR)
-  list(APPEND EP_COMMON_TOOLCHAIN -DCMAKE_AR=${CMAKE_AR})
+  # Ensure using absolute path.
+  find_program(EP_CMAKE_AR ${CMAKE_AR} REQUIRED)
+  list(APPEND EP_COMMON_TOOLCHAIN -DCMAKE_AR=${EP_CMAKE_AR})
 endif()
 
 if(CMAKE_RANLIB)
-  list(APPEND EP_COMMON_TOOLCHAIN -DCMAKE_RANLIB=${CMAKE_RANLIB})
+  # Ensure using absolute path.
+  find_program(EP_CMAKE_RANLIB ${CMAKE_RANLIB} REQUIRED)
+  list(APPEND EP_COMMON_TOOLCHAIN -DCMAKE_RANLIB=${EM_CMAKE_RANLIB})
 endif()
 
 # External projects are still able to override the following declarations.
```

But I still get the following error:

```text
[34/34] Linking C executable brotli.js
FAILED: brotli.js 
: && /bin/emcc -Qunused-arguments -sUSE_ZLIB=1 -sSIDE_MODULE=1 -fPIC -fexceptions -fPIC -g -Werror -O0 -ggdb   -Wno-error  CMakeFiles/brotli.dir/c/tools/brotli.c.o -o brotli.js  libbrotlienc-static.a  libbrotlidec-static.a  libbrotlicommon-static.a && :
emcc: error: output suffix .js requested, but wasm side modules are just wasm files; emitting only a .wasm, no .js [-Wemcc] [-Werror]
ninja: build stopped: subcommand failed.
```

Do you know how to solve this?

joemarshall replied:
This one was debug only - SIDE_MODULE shouldn't be set on executable builds. Should be fixed now.

kou commented on python/CMakeLists.txt:
Could you move Python related changes to #37822?

joemarshall replied:
done

kou reviewed (commented):
I can't build successfully yet.

I need at least the following `Threads::Threads` related changes:

```diff
diff --git a/cpp/CMakeLists.txt b/cpp/CMakeLists.txt
index f2906b960..d7ab9c2fd 100644
--- a/cpp/CMakeLists.txt
+++ b/cpp/CMakeLists.txt
@@ -960,7 +960,10 @@ endif()
 
 set(ARROW_SYSTEM_LINK_LIBS)
 
-list(APPEND ARROW_SYSTEM_LINK_LIBS Threads::Threads)
+list(APPEND ARROW_SYSTEM_LINK_LIBS)
+if(ARROW_ENABLE_THREADING)
+  list(APPEND ARROW_SYSTEM_LINK_LIBS Threads::Threads)
+endif()
 if(CMAKE_THREAD_LIBS_INIT)
   string(APPEND ARROW_PC_LIBS_PRIVATE " ${CMAKE_THREAD_LIBS_INIT}")
 endif()
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index b50f3f1f1..9cd6e6727 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -1040,8 +1044,10 @@ endmacro()
 # ----------------------------------------------------------------------
 # Find pthreads
 
-set(THREADS_PREFER_PTHREAD_FLAG ON)
-find_package(Threads REQUIRED)
+if(ARROW_ENABLE_THREADING)
+  set(THREADS_PREFER_PTHREAD_FLAG ON)
+  find_package(Threads REQUIRED)
+endif()
 
 # ----------------------------------------------------------------------
 # Add Boost dependencies (code adapted from Apache Kudu)
@@ -4494,7 +4500,7 @@ macro(build_orc)
     list(APPEND ORC_LINK_LIBRARIES absl::log_internal_check_op)
   endif()
   if(NOT MSVC)
-    if(NOT APPLE)
+    if(ARROW_ENABLE_THREADING AND NOT APPLE)
       list(APPEND ORC_LINK_LIBRARIES Threads::Threads)
     endif()
     list(APPEND ORC_LINK_LIBRARIES ${CMAKE_DL_LIBS})
```

joemarshall commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -258,6 +258,21 @@ macro(resolve_dependency DEPENDENCY_NAME)
     set(ARG_IS_RUNTIME_DEPENDENCY TRUE)
   endif()
 
+  # ensure zlib is built with -fpic
+  # and make sure that the build finds the version in Emscripten ports
+  # - n.b. the actual linking happens because -sUSE_ZLIB=1 is
+  # set in the compiler variables, but cmake expects
+  # it to exist at configuration time if we aren't building it as
+  # bundled. We need to do this for all packages
+  # not just zlib as some depend on zlib, but we don't rebuild
+  # if it exists already
+  if(CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
+    if(NOT EXISTS ${EMSCRIPTEN_SYSROOT}/lib/wasm32-emscripten/pic/libz.a)
+      execute_process(COMMAND embuilder --pic --force build zlib)
+    endif()
+    set(ZLIB_LIBRARY ${EMSCRIPTEN_SYSROOT}/lib/wasm32-emscripten/pic/libz.a)
+  endif()
```
good idea, implementing it now.

joemarshall commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1365,16 +1398,51 @@ macro(build_brotli)
-                                       ${BROTLI_BUILD_BYPRODUCTS}
-                      CMAKE_ARGS ${BROTLI_CMAKE_ARGS}
-                      STEP_TARGETS headers_copy)
+  set(BROTLI_INSTALL_OVERRIDE)
```
Leftover code, sorry, gone now.

joemarshall commented:
I've done those various review fixes, and fixed the debug build now (here at least). Should build fine with emsdk 3.1.45.

You're right you need `emcmake cmake` by the way, docs fixed.

kou reviewed (commented):
@joemarshall Thanks. Could you also check unresolved comments? You can view them in https://github.com/apache/arrow/pull/37821/files .

Do you really not need thread related changes as I described at https://github.com/apache/arrow/pull/37821#pullrequestreview-1649733026 ?

joemarshall commented on cpp/cmake_modules/EmscriptenOverrides.cmake:
```diff
@@ -0,0 +1,25 @@
+# Force some variables for Emscripten
+# to disable things that won't work there
+
+# # override default in Emscripten which is to not use shared libs
+set_property(GLOBAL PROPERTY TARGET_SUPPORTS_SHARED_LIBS TRUE)
```
nope - there's a few external projects that won't build static libs only and fail to build if you don't enable shared libs (they try and build things twice to the same location). I can't remember which ones, but this was the only way I could workaround that.

joemarshall commented on cpp/CMakePresets.json:
```diff
@@ -368,6 +421,16 @@
+      "inherits": [
+        "features-emscripten",
+        "base-release",
+        "features-python"
```
it is in as
```
"features-emscripten",
"base-release"
```
and 
```
"features-emscripten",
"base-debug"
```
It needs to be in that order because things from base need overriding (base sets 'ARROW_BUILD_STATIC' off.)

joemarshall commented:
@kou I've gone through everything now. I don't have those threading changes in by the way - emscripten has a stub for pthreads, so it should build okay.

jorisvandenbossche commented:
The AMR UBuntu 20.04 C++ build failure is unrelated (new release, and ther are no aarch64 wheels yet: https://sourceforge.net/p/ruamel-yaml-clib/tickets/25/, https://sourceforge.net/p/ruamel-yaml/tickets/479/)

kou commented:
> I don't have those threading changes in by the way - emscripten has a stub for pthreads, so it should build okay.

OK. I checked CMake's log again, `find_package(Threads REQUIRED)` failed with another problem (`pic/` doesn't exist for `libz.a` path):

```text
        FAILED: CMakeFiles/cmTC_7a54a.dir/CheckForPthreads.c.o 
        /bin/emcc   -Qunused-arguments -sUSE_ZLIB=1  -fPIC -fexceptions  -std=gnu11 -fPIE -MD -MT CMakeFiles/cmTC_7a54a.dir/CheckForPthreads.c.o -MF 
CMakeFiles/cmTC_7a54a.dir/CheckForPthreads.c.o.d -o CMakeFiles/cmTC_7a54a.dir/CheckForPthreads.c.o -c /home/kou/work/cpp/arrow.kou/cpp.build.emscripten/CMakeFiles/CMakeScratch/TryCompile-jKLYqN/CheckForPthreads.c
        Traceback (most recent call last):
          File "/usr/share/emscripten/emcc.py", line 3947, in <module>
            sys.exit(main(sys.argv))
                     ^^^^^^^^^^^^^^
          File "/usr/lib/python3.11/contextlib.py", line 81, in inner
            return func(*args, **kwds)
                   ^^^^^^^^^^^^^^^^^^^
          File "/usr/share/emscripten/emcc.py", line 3940, in main
            ret = run(args)
                  ^^^^^^^^^
          File "/usr/share/emscripten/emcc.py", line 1151, in run
            linker_inputs = phase_compile_inputs(options, state, newargs, input_files)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/usr/lib/python3.11/contextlib.py", line 81, in inner
            return func(*args, **kwds)
                   ^^^^^^^^^^^^^^^^^^^
          File "/usr/share/emscripten/emcc.py", line 2677, in phase_compile_inputs
            compile_source_file(i, input_file)
          File "/usr/share/emscripten/emcc.py", line 2657, in compile_source_file
            cmd = get_clang_command(input_file)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/usr/share/emscripten/emcc.py", line 2598, in get_clang_command
            return get_compiler(src_file) + get_cflags(state.orig_args) + compile_args + [src_file]
                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/usr/share/emscripten/emcc.py", line 903, in get_cflags
            ports.add_cflags(cflags, settings)
          File "/usr/share/emscripten/tools/ports/__init__.py", line 364, in add_cflags
            port.get(Ports, settings, shared)
          File "/usr/share/emscripten/tools/ports/zlib.py", line 46, in get
            return [shared.Cache.get_lib('libz.a', create, what='port')]
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/usr/share/emscripten/tools/cache.py", line 139, in get_lib
            return self.get(name, *args, **kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "/usr/share/emscripten/tools/cache.py", line 154, in get
            raise Exception(f'FROZEN_CACHE is set, but cache file is missing: "{shortname}" (in cache root path "{self.dirname}")')
        Exception: FROZEN_CACHE is set, but cache file is missing: "sysroot/lib/wasm32-emscripten/libz.a" (in cache root path "/usr/share/emscripten/
cache")
        ninja: build stopped: subcommand failed.
```

Anyway, let's remove `Threads::Threads` dependency because it's needless for `ARROW_ENABLE_THREADING=OFF`: https://github.com/apache/arrow/issues/38055

kou commented:
Could you apply the following changes to add a test for this case? (Or can I push it?)

```diff
diff --git a/ci/docker/ubuntu-22.04-cpp.dockerfile b/ci/docker/ubuntu-22.04-cpp.dockerfile
index fffafe2b0..bf8f370be 100644
--- a/ci/docker/ubuntu-22.04-cpp.dockerfile
+++ b/ci/docker/ubuntu-22.04-cpp.dockerfile
@@ -69,6 +69,7 @@ RUN apt-get update -y -q && \
         ccache \
         cmake \
         curl \
+        emscripten \
         gdb \
         git \
         libbenchmark-dev \
@@ -149,6 +150,10 @@ RUN if [ "${gcc_version}" = "" ]; then \
       update-alternatives --set c++ /usr/bin/g++; \
     fi
 
+RUN EM_CACHE=/usr/share/emscripten/cache \
+    EM_FROZEN_CACHE= \
+      embuilder --pic build zlib
+
 COPY ci/scripts/install_minio.sh /arrow/ci/scripts/
 RUN /arrow/ci/scripts/install_minio.sh latest /usr/local
 
diff --git a/ci/scripts/cpp_build.sh b/ci/scripts/cpp_build.sh
index 1f5596e2a..8e13b0590 100755
--- a/ci/scripts/cpp_build.sh
+++ b/ci/scripts/cpp_build.sh
@@ -81,111 +81,125 @@ esac
 mkdir -p ${build_dir}
 pushd ${build_dir}
 
-cmake \
-  -Dabsl_SOURCE=${absl_SOURCE:-} \
-  -DARROW_ACERO=${ARROW_ACERO:-OFF} \
-  -DARROW_AZURE=${ARROW_AZURE:-OFF} \
-  -DARROW_BOOST_USE_SHARED=${ARROW_BOOST_USE_SHARED:-ON} \
-  -DARROW_BUILD_BENCHMARKS_REFERENCE=${ARROW_BUILD_BENCHMARKS:-OFF} \
-  -DARROW_BUILD_BENCHMARKS=${ARROW_BUILD_BENCHMARKS:-OFF} \
-  -DARROW_BUILD_EXAMPLES=${ARROW_BUILD_EXAMPLES:-OFF} \
-  -DARROW_BUILD_INTEGRATION=${ARROW_BUILD_INTEGRATION:-OFF} \
-  -DARROW_BUILD_SHARED=${ARROW_BUILD_SHARED:-ON} \
-  -DARROW_BUILD_STATIC=${ARROW_BUILD_STATIC:-ON} \
-  -DARROW_BUILD_TESTS=${ARROW_BUILD_TESTS:-OFF} \
-  -DARROW_BUILD_UTILITIES=${ARROW_BUILD_UTILITIES:-ON} \
-  -DARROW_COMPUTE=${ARROW_COMPUTE:-ON} \
-  -DARROW_CSV=${ARROW_CSV:-ON} \
-  -DARROW_CUDA=${ARROW_CUDA:-OFF} \
-  -DARROW_CXXFLAGS=${ARROW_CXXFLAGS:-} \
-  -DARROW_CXX_FLAGS_DEBUG="${ARROW_CXX_FLAGS_DEBUG:-}" \
-  -DARROW_CXX_FLAGS_RELEASE="${ARROW_CXX_FLAGS_RELEASE:-}" \
-  -DARROW_CXX_FLAGS_RELWITHDEBINFO="${ARROW_CXX_FLAGS_RELWITHDEBINFO:-}" \
-  -DARROW_C_FLAGS_DEBUG="${ARROW_C_FLAGS_DEBUG:-}" \
-  -DARROW_C_FLAGS_RELEASE="${ARROW_C_FLAGS_RELEASE:-}" \
-  -DARROW_C_FLAGS_RELWITHDEBINFO="${ARROW_C_FLAGS_RELWITHDEBINFO:-}" \
-  -DARROW_DATASET=${ARROW_DATASET:-OFF} \
-  -DARROW_DEPENDENCY_SOURCE=${ARROW_DEPENDENCY_SOURCE:-AUTO} \
-  -DARROW_ENABLE_THREADING=${ARROW_ENABLE_THREADING:-ON} \
-  -DARROW_ENABLE_TIMING_TESTS=${ARROW_ENABLE_TIMING_TESTS:-ON} \
-  -DARROW_EXTRA_ERROR_CONTEXT=${ARROW_EXTRA_ERROR_CONTEXT:-OFF} \
-  -DARROW_FILESYSTEM=${ARROW_FILESYSTEM:-ON} \
-  -DARROW_FLIGHT=${ARROW_FLIGHT:-OFF} \
-  -DARROW_FLIGHT_SQL=${ARROW_FLIGHT_SQL:-OFF} \
-  -DARROW_FUZZING=${ARROW_FUZZING:-OFF} \
-  -DARROW_GANDIVA_PC_CXX_FLAGS=${ARROW_GANDIVA_PC_CXX_FLAGS:-} \
-  -DARROW_GANDIVA=${ARROW_GANDIVA:-OFF} \
-  -DARROW_GCS=${ARROW_GCS:-OFF} \
-  -DARROW_HDFS=${ARROW_HDFS:-ON} \
-  -DARROW_INSTALL_NAME_RPATH=${ARROW_INSTALL_NAME_RPATH:-ON} \
-  -DARROW_JEMALLOC=${ARROW_JEMALLOC:-ON} \
-  -DARROW_JSON=${ARROW_JSON:-ON} \
-  -DARROW_LARGE_MEMORY_TESTS=${ARROW_LARGE_MEMORY_TESTS:-OFF} \
-  -DARROW_MIMALLOC=${ARROW_MIMALLOC:-OFF} \
-  -DARROW_NO_DEPRECATED_API=${ARROW_NO_DEPRECATED_API:-OFF} \
-  -DARROW_ORC=${ARROW_ORC:-OFF} \
-  -DARROW_PARQUET=${ARROW_PARQUET:-OFF} \
-  -DARROW_RUNTIME_SIMD_LEVEL=${ARROW_RUNTIME_SIMD_LEVEL:-MAX} \
-  -DARROW_S3=${ARROW_S3:-OFF} \
-  -DARROW_SIMD_LEVEL=${ARROW_SIMD_LEVEL:-DEFAULT} \
-  -DARROW_SKYHOOK=${ARROW_SKYHOOK:-OFF} \
-  -DARROW_SUBSTRAIT=${ARROW_SUBSTRAIT:-OFF} \
-  -DARROW_TEST_LINKAGE=${ARROW_TEST_LINKAGE:-shared} \
-  -DARROW_TEST_MEMCHECK=${ARROW_TEST_MEMCHECK:-OFF} \
-  -DARROW_USE_ASAN=${ARROW_USE_ASAN:-OFF} \
-  -DARROW_USE_CCACHE=${ARROW_USE_CCACHE:-ON} \
-  -DARROW_USE_GLOG=${ARROW_USE_GLOG:-OFF} \
-  -DARROW_USE_LD_GOLD=${ARROW_USE_LD_GOLD:-OFF} \
-  -DARROW_USE_PRECOMPILED_HEADERS=${ARROW_USE_PRECOMPILED_HEADERS:-OFF} \
-  -DARROW_USE_STATIC_CRT=${ARROW_USE_STATIC_CRT:-OFF} \
-  -DARROW_USE_TSAN=${ARROW_USE_TSAN:-OFF} \
-  -DARROW_USE_UBSAN=${ARROW_USE_UBSAN:-OFF} \
-  -DARROW_VERBOSE_THIRDPARTY_BUILD=${ARROW_VERBOSE_THIRDPARTY_BUILD:-OFF} \
-  -DARROW_WITH_BROTLI=${ARROW_WITH_BROTLI:-OFF} \
-  -DARROW_WITH_BZ2=${ARROW_WITH_BZ2:-OFF} \
-  -DARROW_WITH_LZ4=${ARROW_WITH_LZ4:-OFF} \
-  -DARROW_WITH_OPENTELEMETRY=${ARROW_WITH_OPENTELEMETRY:-OFF} \
-  -DARROW_WITH_MUSL=${ARROW_WITH_MUSL:-OFF} \
-  -DARROW_WITH_SNAPPY=${ARROW_WITH_SNAPPY:-OFF} \
-  -DARROW_WITH_UCX=${ARROW_WITH_UCX:-OFF} \
-  -DARROW_WITH_UTF8PROC=${ARROW_WITH_UTF8PROC:-ON} \
-  -DARROW_WITH_ZLIB=${ARROW_WITH_ZLIB:-OFF} \
-  -DARROW_WITH_ZSTD=${ARROW_WITH_ZSTD:-OFF} \
-  -DAWSSDK_SOURCE=${AWSSDK_SOURCE:-} \
-  -DAzure_SOURCE=${Azure_SOURCE:-} \
-  -Dbenchmark_SOURCE=${benchmark_SOURCE:-} \
-  -DBOOST_SOURCE=${BOOST_SOURCE:-} \
-  -DBrotli_SOURCE=${Brotli_SOURCE:-} \
-  -DBUILD_WARNING_LEVEL=${BUILD_WARNING_LEVEL:-CHECKIN} \
-  -Dc-ares_SOURCE=${cares_SOURCE:-} \
-  -DCMAKE_BUILD_TYPE=${ARROW_BUILD_TYPE:-debug} \
-  -DCMAKE_VERBOSE_MAKEFILE=${CMAKE_VERBOSE_MAKEFILE:-OFF} \
-  -DCMAKE_C_FLAGS="${CFLAGS:-}" \
-  -DCMAKE_CXX_FLAGS="${CXXFLAGS:-}" \
-  -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD:-17}" \
-  -DCMAKE_INSTALL_LIBDIR=${CMAKE_INSTALL_LIBDIR:-lib} \
-  -DCMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX:-${ARROW_HOME}} \
-  -DCMAKE_UNITY_BUILD=${CMAKE_UNITY_BUILD:-OFF} \
-  -Dgflags_SOURCE=${gflags_SOURCE:-} \
-  -Dgoogle_cloud_cpp_storage_SOURCE=${google_cloud_cpp_storage_SOURCE:-} \
-  -DgRPC_SOURCE=${gRPC_SOURCE:-} \
-  -DGTest_SOURCE=${GTest_SOURCE:-} \
-  -Dlz4_SOURCE=${lz4_SOURCE:-} \
-  -DORC_SOURCE=${ORC_SOURCE:-} \
-  -DPARQUET_BUILD_EXAMPLES=${PARQUET_BUILD_EXAMPLES:-OFF} \
-  -DPARQUET_BUILD_EXECUTABLES=${PARQUET_BUILD_EXECUTABLES:-OFF} \
-  -DPARQUET_REQUIRE_ENCRYPTION=${PARQUET_REQUIRE_ENCRYPTION:-ON} \
-  -DProtobuf_SOURCE=${Protobuf_SOURCE:-} \
-  -DRapidJSON_SOURCE=${RapidJSON_SOURCE:-} \
-  -Dre2_SOURCE=${re2_SOURCE:-} \
-  -DSnappy_SOURCE=${Snappy_SOURCE:-} \
-  -DThrift_SOURCE=${Thrift_SOURCE:-} \
-  -Dutf8proc_SOURCE=${utf8proc_SOURCE:-} \
-  -Dzstd_SOURCE=${zstd_SOURCE:-} \
-  -Dxsimd_SOURCE=${xsimd_SOURCE:-} \
-  -G "${CMAKE_GENERATOR:-Ninja}" \
-  ${ARROW_CMAKE_ARGS} \
-  ${source_dir}
+if [ "${ARROW_EMSCRIPTEN:-OFF}" = "ON" ]; then
+  emcmake cmake \
+    --preset=ninja-${ARROW_BUILD_TYPE:-debug}-emscripten \
+    -DCMAKE_VERBOSE_MAKEFILE=${CMAKE_VERBOSE_MAKEFILE:-OFF} \
+    -DCMAKE_C_FLAGS="${CFLAGS:-}" \
+    -DCMAKE_CXX_FLAGS="${CXXFLAGS:-}" \
+    -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD:-17}" \
+    -DCMAKE_INSTALL_LIBDIR=${CMAKE_INSTALL_LIBDIR:-lib} \
+    -DCMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX:-${ARROW_HOME}} \
+    -DCMAKE_UNITY_BUILD=${CMAKE_UNITY_BUILD:-OFF} \
+    ${ARROW_CMAKE_ARGS} \
+    ${source_dir}
+else
+  cmake \
+    -Dabsl_SOURCE=${absl_SOURCE:-} \
+    -DARROW_ACERO=${ARROW_ACERO:-OFF} \
+    -DARROW_AZURE=${ARROW_AZURE:-OFF} \
+    -DARROW_BOOST_USE_SHARED=${ARROW_BOOST_USE_SHARED:-ON} \
+    -DARROW_BUILD_BENCHMARKS_REFERENCE=${ARROW_BUILD_BENCHMARKS:-OFF} \
+    -DARROW_BUILD_BENCHMARKS=${ARROW_BUILD_BENCHMARKS:-OFF} \
+    -DARROW_BUILD_EXAMPLES=${ARROW_BUILD_EXAMPLES:-OFF} \
+    -DARROW_BUILD_INTEGRATION=${ARROW_BUILD_INTEGRATION:-OFF} \
+    -DARROW_BUILD_SHARED=${ARROW_BUILD_SHARED:-ON} \
+    -DARROW_BUILD_STATIC=${ARROW_BUILD_STATIC:-ON} \
+    -DARROW_BUILD_TESTS=${ARROW_BUILD_TESTS:-OFF} \
+    -DARROW_BUILD_UTILITIES=${ARROW_BUILD_UTILITIES:-ON} \
+    -DARROW_COMPUTE=${ARROW_COMPUTE:-ON} \
+    -DARROW_CSV=${ARROW_CSV:-ON} \
+    -DARROW_CUDA=${ARROW_CUDA:-OFF} \
+    -DARROW_CXXFLAGS=${ARROW_CXXFLAGS:-} \
+    -DARROW_CXX_FLAGS_DEBUG="${ARROW_CXX_FLAGS_DEBUG:-}" \
+    -DARROW_CXX_FLAGS_RELEASE="${ARROW_CXX_FLAGS_RELEASE:-}" \
+    -DARROW_CXX_FLAGS_RELWITHDEBINFO="${ARROW_CXX_FLAGS_RELWITHDEBINFO:-}" \
+    -DARROW_C_FLAGS_DEBUG="${ARROW_C_FLAGS_DEBUG:-}" \
+    -DARROW_C_FLAGS_RELEASE="${ARROW_C_FLAGS_RELEASE:-}" \
+    -DARROW_C_FLAGS_RELWITHDEBINFO="${ARROW_C_FLAGS_RELWITHDEBINFO:-}" \
+    -DARROW_DATASET=${ARROW_DATASET:-OFF} \
+    -DARROW_DEPENDENCY_SOURCE=${ARROW_DEPENDENCY_SOURCE:-AUTO} \
+    -DARROW_ENABLE_THREADING=${ARROW_ENABLE_THREADING:-ON} \
+    -DARROW_ENABLE_TIMING_TESTS=${ARROW_ENABLE_TIMING_TESTS:-ON} \
+    -DARROW_EXTRA_ERROR_CONTEXT=${ARROW_EXTRA_ERROR_CONTEXT:-OFF} \
+    -DARROW_FILESYSTEM=${ARROW_FILESYSTEM:-ON} \
+    -DARROW_FLIGHT=${ARROW_FLIGHT:-OFF} \
+    -DARROW_FLIGHT_SQL=${ARROW_FLIGHT_SQL:-OFF} \
+    -DARROW_FUZZING=${ARROW_FUZZING:-OFF} \
+    -DARROW_GANDIVA_PC_CXX_FLAGS=${ARROW_GANDIVA_PC_CXX_FLAGS:-} \
+    -DARROW_GANDIVA=${ARROW_GANDIVA:-OFF} \
+    -DARROW_GCS=${ARROW_GCS:-OFF} \
+    -DARROW_HDFS=${ARROW_HDFS:-ON} \
+    -DARROW_INSTALL_NAME_RPATH=${ARROW_INSTALL_NAME_RPATH:-ON} \
+    -DARROW_JEMALLOC=${ARROW_JEMALLOC:-ON} \
+    -DARROW_JSON=${ARROW_JSON:-ON} \
+    -DARROW_LARGE_MEMORY_TESTS=${ARROW_LARGE_MEMORY_TESTS:-OFF} \
+    -DARROW_MIMALLOC=${ARROW_MIMALLOC:-OFF} \
+    -DARROW_NO_DEPRECATED_API=${ARROW_NO_DEPRECATED_API:-OFF} \
+    -DARROW_ORC=${ARROW_ORC:-OFF} \
+    -DARROW_PARQUET=${ARROW_PARQUET:-OFF} \
+    -DARROW_RUNTIME_SIMD_LEVEL=${ARROW_RUNTIME_SIMD_LEVEL:-MAX} \
+    -DARROW_S3=${ARROW_S3:-OFF} \
+    -DARROW_SIMD_LEVEL=${ARROW_SIMD_LEVEL:-DEFAULT} \
+    -DARROW_SKYHOOK=${ARROW_SKYHOOK:-OFF} \
+    -DARROW_SUBSTRAIT=${ARROW_SUBSTRAIT:-OFF} \
+    -DARROW_TEST_LINKAGE=${ARROW_TEST_LINKAGE:-shared} \
+    -DARROW_TEST_MEMCHECK=${ARROW_TEST_MEMCHECK:-OFF} \
+    -DARROW_USE_ASAN=${ARROW_USE_ASAN:-OFF} \
+    -DARROW_USE_CCACHE=${ARROW_USE_CCACHE:-ON} \
+    -DARROW_USE_GLOG=${ARROW_USE_GLOG:-OFF} \
+    -DARROW_USE_LD_GOLD=${ARROW_USE_LD_GOLD:-OFF} \
+    -DARROW_USE_PRECOMPILED_HEADERS=${ARROW_USE_PRECOMPILED_HEADERS:-OFF} \
+    -DARROW_USE_STATIC_CRT=${ARROW_USE_STATIC_CRT:-OFF} \
+    -DARROW_USE_TSAN=${ARROW_USE_TSAN:-OFF} \
+    -DARROW_USE_UBSAN=${ARROW_USE_UBSAN:-OFF} \
+    -DARROW_VERBOSE_THIRDPARTY_BUILD=${ARROW_VERBOSE_THIRDPARTY_BUILD:-OFF} \
+    -DARROW_WITH_BROTLI=${ARROW_WITH_BROTLI:-OFF} \
+    -DARROW_WITH_BZ2=${ARROW_WITH_BZ2:-OFF} \
+    -DARROW_WITH_LZ4=${ARROW_WITH_LZ4:-OFF} \
+    -DARROW_WITH_OPENTELEMETRY=${ARROW_WITH_OPENTELEMETRY:-OFF} \
+    -DARROW_WITH_MUSL=${ARROW_WITH_MUSL:-OFF} \
+    -DARROW_WITH_SNAPPY=${ARROW_WITH_SNAPPY:-OFF} \
+    -DARROW_WITH_UCX=${ARROW_WITH_UCX:-OFF} \
+    -DARROW_WITH_UTF8PROC=${ARROW_WITH_UTF8PROC:-ON} \
+    -DARROW_WITH_ZLIB=${ARROW_WITH_ZLIB:-OFF} \
+    -DARROW_WITH_ZSTD=${ARROW_WITH_ZSTD:-OFF} \
+    -DAWSSDK_SOURCE=${AWSSDK_SOURCE:-} \
+    -DAzure_SOURCE=${Azure_SOURCE:-} \
+    -Dbenchmark_SOURCE=${benchmark_SOURCE:-} \
+    -DBOOST_SOURCE=${BOOST_SOURCE:-} \
+    -DBrotli_SOURCE=${Brotli_SOURCE:-} \
+    -DBUILD_WARNING_LEVEL=${BUILD_WARNING_LEVEL:-CHECKIN} \
+    -Dc-ares_SOURCE=${cares_SOURCE:-} \
+    -DCMAKE_BUILD_TYPE=${ARROW_BUILD_TYPE:-debug} \
+    -DCMAKE_VERBOSE_MAKEFILE=${CMAKE_VERBOSE_MAKEFILE:-OFF} \
+    -DCMAKE_C_FLAGS="${CFLAGS:-}" \
+    -DCMAKE_CXX_FLAGS="${CXXFLAGS:-}" \
+    -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD:-17}" \
+    -DCMAKE_INSTALL_LIBDIR=${CMAKE_INSTALL_LIBDIR:-lib} \
+    -DCMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX:-${ARROW_HOME}} \
+    -DCMAKE_UNITY_BUILD=${CMAKE_UNITY_BUILD:-OFF} \
+    -Dgflags_SOURCE=${gflags_SOURCE:-} \
+    -Dgoogle_cloud_cpp_storage_SOURCE=${google_cloud_cpp_storage_SOURCE:-} \
+    -DgRPC_SOURCE=${gRPC_SOURCE:-} \
+    -DGTest_SOURCE=${GTest_SOURCE:-} \
+    -Dlz4_SOURCE=${lz4_SOURCE:-} \
+    -DORC_SOURCE=${ORC_SOURCE:-} \
+    -DPARQUET_BUILD_EXAMPLES=${PARQUET_BUILD_EXAMPLES:-OFF} \
+    -DPARQUET_BUILD_EXECUTABLES=${PARQUET_BUILD_EXECUTABLES:-OFF} \
+    -DPARQUET_REQUIRE_ENCRYPTION=${PARQUET_REQUIRE_ENCRYPTION:-ON} \
+    -DProtobuf_SOURCE=${Protobuf_SOURCE:-} \
+    -DRapidJSON_SOURCE=${RapidJSON_SOURCE:-} \
+    -Dre2_SOURCE=${re2_SOURCE:-} \
+    -DSnappy_SOURCE=${Snappy_SOURCE:-} \
+    -DThrift_SOURCE=${Thrift_SOURCE:-} \
+    -Dutf8proc_SOURCE=${utf8proc_SOURCE:-} \
+    -Dzstd_SOURCE=${zstd_SOURCE:-} \
+    -Dxsimd_SOURCE=${xsimd_SOURCE:-} \
+    -G "${CMAKE_GENERATOR:-Ninja}" \
+    ${ARROW_CMAKE_ARGS} \
+    ${source_dir}
+fi
 
 export CMAKE_BUILD_PARALLEL_LEVEL=${CMAKE_BUILD_PARALLEL_LEVEL:-$[${n_jobs} + 1]}
 time cmake --build . --target install
diff --git a/dev/tasks/tasks.yml b/dev/tasks/tasks.yml
index 859ff8ddb..b8b0d9ed2 100644
--- a/dev/tasks/tasks.yml
+++ b/dev/tasks/tasks.yml
@@ -1255,6 +1255,14 @@ tasks:
       flags: "-e ARROW_CSV=ON -e ARROW_PARQUET=ON"
       image: ubuntu-cpp-minimal
 
+  test-ubuntu-22.04-cpp-emscripten:
+    ci: github
+    template: docker-tests/github.linux.yml
+    params:
+      env:
+        UBUNTU: 22.04
+      image: ubuntu-cpp-emscripten
+
 {% for python_version in ["3.8", "3.9", "3.10", "3.11"] %}
   test-conda-python-{{ python_version }}:
     ci: github
diff --git a/docker-compose.yml b/docker-compose.yml
index 62e5aee0a..ea3b19281 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -158,6 +158,7 @@ x-hierarchy:
   - ubuntu-csharp
   - ubuntu-cpp-sanitizer
   - ubuntu-cpp-thread-sanitizer
+  - ubuntu-cpp-emscripten
   - ubuntu-r-sanitizer
   - ubuntu-r-valgrind
   - ubuntu-swift
@@ -649,6 +650,30 @@ services:
       ARROW_USE_TSAN: "ON"
     command: *cpp-command
 
+  ubuntu-cpp-emscripten:
+    # Usage:
+    #   docker-compose build ubuntu-cpp-emscripten
+    #   docker-compose run --rm ubuntu-cpp-emscripten
+    # Parameters:
+    #   ARCH: amd64, arm64v8, ...
+    #   UBUNTU: 22.04
+    image: ${REPO}:${ARCH}-ubuntu-${UBUNTU}-cpp
+    build:
+      context: .
+      dockerfile: ci/docker/ubuntu-${UBUNTU}-cpp.dockerfile
+      cache_from:
+        - ${REPO}:${ARCH}-ubuntu-${UBUNTU}-cpp
+      args:
+        arch: ${ARCH}
+        clang_tools: ${CLANG_TOOLS}
+        llvm: ${LLVM}
+    shm_size: *shm-size
+    volumes: *ubuntu-volumes
+    environment:
+      <<: [*common, *ccache, *sccache, *cpp]
+      ARROW_EMSCRIPTEN: "ON"
+    command: *cpp-command
+
   fedora-cpp:
     # Usage:
     #   docker-compose build fedora-cpp
```

kou commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -700,3 +702,30 @@ if(MSVC)
+  # size_t is 32 bit in emscripten wasm32 - ignore conversion errors
+  # deprecated-literal-operator error is thrown in datetime (vendored lib in arrow)
+  set(CMAKE_CXX_FLAGS
+      "${CMAKE_CXX_FLAGS} -sUSE_ZLIB=1 -fPIC -fexceptions -Wno-error=shorten-64-to-32 -Wno-error=deprecated-literal-operator"
```
#38090 for `shorten-64-to-32`.

kou replied:
@joemarshall Could you check this? Or can I push some commits including this to this branch?

joemarshall replied:
oh, forgot to tick this off, I did this.

joemarshall commented:
Ah, I see why it isn't working - you need to install emscripten with the emsdk, instructions here:
https://emscripten.org/docs/getting_started/downloads.html 
I would set the version to 3.1.45 for now, because we know it builds on that, unless you want to set it to latest so as to keep up with emscripten releases (but breaking CI if the emscripten build breaks).

The version from ubuntu is 3.1.5, pyodide builds with 3.1.45. I suspect that it is using an older compiler etc. hence the weird compiler errors I don't see here.

I couldn't quite work out how to do that in the dockerfile, so if you could do that it would be great:

Basically, you need to:
1) clone the git and run `./emsdk install 3.1.45`, `./emsdk activate 3.1.45`.
2) Call 'source emsdk/emsdk_env.sh` to load that emsdk into your environment before any build commands (I guess that can go in profile or somewhere).

kou commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-emscripten

raulcd commented:
This is marked as 14.0.0. I am removing the milestone. Please mark the related issue with the `Priority: blocker` label if it must be a blocker for the release.

joemarshall commented:
@kou - I did those changes to the dockerfile, so now if you run

`UBUNTU=22.04 archery docker run ubuntu-cpp-emscripten` it builds okay.

jorisvandenbossche commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-emscripten

joemarshall commented:
@jorisvandenbossche Could you submit that build again - it appears to be running out of memory because of:
https://github.com/WebAssembly/binaryen/issues/4261
I've applied a fix which sets `CMAKE_BUILD_PARALLEL_LEVEL` to 1 for debug builds in emscripten.

raulcd commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-emscripten

jorisvandenbossche commented:
The build is still failing:

```
 [371/468] Linking CXX executable debug/arrow-json-integration-test.js
FAILED: debug/arrow-json-integration-test.js 
: && /root/emsdk/upstream/emscripten/em++ -Qunused-arguments -fcolor-diagnostics -sUSE_ZLIB=1 -fPIC -fexceptions -Wno-error=shorten-64-to-32 -Wno-error=deprecated-literal-operator  -Wall -Wextra -Wdocumentation -Wshorten-64-to-32 -Wno-missing-braces -Wno-unused-parameter -Wno-constant-logical-operand -Wno-return-stack-address -Wdate-time -Wno-unknown-warning-option -Wno-pass-failed  -g -Werror -O0 -ggdb  src/arrow/integration/CMakeFiles/arrow-json-integration-test.dir/json_integration_test.cc.o -o debug/arrow-json-integration-test.js  debug/libarrow_testing.a  debug/libarrow.a  brotli_ep/src/brotli_ep-install/lib/libbrotlienc-static.a  brotli_ep/src/brotli_ep-install/lib/libbrotlidec-static.a  brotli_ep/src/brotli_ep-install/lib/libbrotlicommon-static.a  orc_ep-install/lib/liborc.a  protobuf_ep-install/lib/libprotobuf.a  utf8proc_ep-install/lib/libutf8proc.a  re2_ep-install/lib/libre2.a  lib/libarrow_gtest_maind.a  lib/libarrow_gtestd.a  lib/libarrow_gmockd.a  gflags_ep-prefix/src/gflags_ep/lib/libgflags_debug.a  lib/libarrow_gtestd.a  lz4_ep-install/lib/liblz4.a  snappy_ep/src/snappy_ep-install/lib/libsnappy.a  /root/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/pic/libz.a  zstd_ep-install/lib/libzstd.a  -lrt && :
em++: error: '/root/emsdk/upstream/bin/wasm-emscripten-finalize -g --dyncalls-i64 --pass-arg=legalize-js-interface-exported-helpers --dwarf debug/arrow-json-integration-test.wasm -o debug/arrow-json-integration-test.wasm --detect-features' failed (received SIGKILL (-9))
ninja: build stopped: subcommand failed.
```

joemarshall commented:
@jorisvandenbossche Right - I split the debug information out of the object files and it seems to take way less memory to build, so crossed fingers this will build now without running out of memory.

jorisvandenbossche commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-emscripten

joemarshall commented:
@jorisvandenbossche @kou 
This appears to build okay now. Right now it isn't running the test suite however. 

I've done the updates to enable tests correctly on emscripten - do you want them in this PR or in a separate one?

kou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -2141,8 +2237,15 @@ function(build_gtest)
-  set(BUILD_STATIC_LIBS OFF)
+  # If we're building static libs for Emscripten, we need to build *everything* as
+  # static libs.
+  if(NOT (CMAKE_SYSTEM_NAME STREQUAL "Emscripten") OR ARROW_BUILD_SHARED)
+    set(BUILD_SHARED_LIBS ON)
+    set(BUILD_STATIC_LIBS OFF)
+  else()
+    set(BUILD_SHARED_LIBS OFF)
+    set(BUILD_STATIC_LIBS ON)
+  endif()
```
We don't need `ARROW_BUILD_SHARED` here:

```suggestion
  if(CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
    set(BUILD_SHARED_LIBS OFF)
    set(BUILD_STATIC_LIBS ON)
  else()
    set(BUILD_SHARED_LIBS ON)
    set(BUILD_STATIC_LIBS OFF)
  endif()
```

kou commented on cpp/src/arrow/compute/kernels/scalar_string_test.cc:
```diff
@@ -1878,17 +1879,29 @@ TYPED_TEST(TestStringKernels, Strptime) {
       Invalid, testing::HasSubstr("Invalid: Failed to parse string: '5/1/2020'"),
       Strptime(ArrayFromJSON(this->type(), input1), options));
+
+  #else
+    GTEST_SKIP()<< "Skipping some strptime tests due to emscripten bug https://github.com/emscripten-core/emscripten/issues/20466";
+  #endif
```
Could you put the following at the beginning instead of `#infdef #else #endif`?

```cpp
  #ifdef EMSCRIPTEN
    GTEST_SKIP()<< "Skipping some strptime tests due to emscripten bug https://github.com/emscripten-core/emscripten/issues/20466";
  #endif
```

kou commented on cpp/src/arrow/compute/kernels/scalar_string_test.cc:
```diff
@@ -1878,17 +1879,29 @@ TYPED_TEST(TestStringKernels, Strptime) {
 TYPED_TEST(TestStringKernels, StrptimeZoneOffset) {
+#ifdef EMSCRIPTEN
+  GTEST_SKIP() << "Emscripten bug https://github.com/emscripten-core/emscripten/issues/20467 ";
+#else
```
We don't need `#else`.

```suggestion
#endif
```

kou commented on cpp/src/arrow/compute/kernels/scalar_string_test.cc:
```diff
@@ -1906,6 +1919,7 @@ TYPED_TEST(TestStringKernels, StrptimeZoneOffset) {
   StrptimeOptions options2("%Y-%m-%dT%H:%M%z", TimeUnit::MICRO, /*error_is_null=*/true);
   this->CheckUnary("strptime", input2, timestamp(TimeUnit::MICRO, "UTC"), output,
                    &options2);
+#endif                   
```
```suggestion
```

kou commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -2159,13 +2162,17 @@ TEST_F(ScalarTemporalTest, StrftimeOtherLocale) {
 TEST_F(ScalarTemporalTest, StrftimeInvalidLocale) {
+#ifdef EMSCRIPTEN
+  GTEST_SKIP() << "Emscripten doesn't build with multiple locales as default";
+#else
```
We don't need `#else`.

```suggestion
#endif
```

kou commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -2159,13 +2162,17 @@ TEST_F(ScalarTemporalTest, StrftimeOtherLocale) {
   EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid,
                                   testing::HasSubstr("Cannot find locale 'non-existent'"),
                                   Strftime(arr, options));
+#endif
```
```suggestion
```

kou commented on cpp/src/arrow/io/file_test.cc:
```diff
@@ -991,6 +1001,8 @@ TEST_F(TestMemoryMappedFile, LARGE_MEMORY_TEST(ReadWriteOver4GbFile)) {
 }
 
 TEST_F(TestMemoryMappedFile, RetainMemoryMapReference) {
+
+
```
Could you revert needless changes?

```suggestion
```

kou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1395,16 +1420,49 @@ macro(build_brotli)
+    set(BROTLI_BUILD_DIR ${CMAKE_CURRENT_BINARY_DIR}/brotli_ep-prefix/src/brotli_ep-build)
+    set(BROTLI_BUILD_LIBS
+        "${BROTLI_BUILD_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}brotlienc-static${CMAKE_STATIC_LIBRARY_SUFFIX}"
+        "${BROTLI_BUILD_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}brotlidec-static${CMAKE_STATIC_LIBRARY_SUFFIX}"
+        "${BROTLI_BUILD_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}brotlicommon-static${CMAKE_STATIC_LIBRARY_SUFFIX}"
+    )
+
+    set(BROTLI_BUILD_INCLUDE_DIR
+        ${CMAKE_CURRENT_BINARY_DIR}/brotli_ep-prefix/src/brotli_ep/c/include/brotli)
+
+    externalproject_add(brotli_ep
+                        ${EP_COMMON_OPTIONS}
+                        URL ${BROTLI_SOURCE_URL}
+                        URL_HASH "SHA256=${ARROW_BROTLI_BUILD_SHA256_CHECKSUM}"
+                        BUILD_BYPRODUCTS "${BROTLI_STATIC_LIBRARY_ENC}"
+                                         "${BROTLI_STATIC_LIBRARY_DEC}"
+                                         "${BROTLI_STATIC_LIBRARY_COMMON}"
+                                         ${BROTLI_BUILD_BYPRODUCTS}
+                        CMAKE_ARGS ${BROTLI_CMAKE_ARGS}
+                        STEP_TARGETS headers_copy
+                        INSTALL_COMMAND "")
+    add_custom_command(TARGET brotli_ep
+                       POST_BUILD
+                       COMMAND ${CMAKE_COMMAND} -E copy_if_different
+                               ${BROTLI_BUILD_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}*${CMAKE_STATIC_LIBRARY_SUFFIX}
+                               ${BROTLI_LIB_DIR}
+                       COMMAND ${CMAKE_COMMAND} -E copy_directory
+                               ${BROTLI_BUILD_INCLUDE_DIR} ${BROTLI_INCLUDE_DIR}/brotli)
+  else() # not emscripten - just behave as normal
+    externalproject_add(brotli_ep
+                        ${EP_COMMON_OPTIONS}
+                        URL ${BROTLI_SOURCE_URL}
+                        URL_HASH "SHA256=${ARROW_BROTLI_BUILD_SHA256_CHECKSUM}"
+                        BUILD_BYPRODUCTS "${BROTLI_STATIC_LIBRARY_ENC}"
+                                         "${BROTLI_STATIC_LIBRARY_DEC}"
+                                         "${BROTLI_STATIC_LIBRARY_COMMON}"
+                                         ${BROTLI_BUILD_BYPRODUCTS}
+                        CMAKE_ARGS ${BROTLI_CMAKE_ARGS}
+                        STEP_TARGETS headers_copy)
+  endif()
```
Could you reuse `externalproject_add()` call?

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index f71f09cbe..f51ec56fc 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -1420,9 +1420,26 @@ macro(build_brotli)
   )
   set(BROTLI_CMAKE_ARGS ${EP_COMMON_CMAKE_ARGS} "-DCMAKE_INSTALL_PREFIX=${BROTLI_PREFIX}")
 
+  set(BROTLI_EP_OPTIONS)
   if(CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
-    # cmake install is disabled for brotli on emscripten, so we have
-    # to manually copy the libraries to our install directory
+    # "cmake install" is disabled for Brotli on Emscripten, so the
+    # default INSTALL_COMMAND is failed. We need to disable the default
+    # INSTALL_COMMAND.
+    list(APPEND BROTLI_EP_OPTIONS INSTALL_COMMAND ${CMAKE_COMMAND} -E true)
+  endif()
+  externalproject_add(brotli_ep
+                      ${EP_COMMON_OPTIONS}
+                      URL ${BROTLI_SOURCE_URL}
+                      URL_HASH "SHA256=${ARROW_BROTLI_BUILD_SHA256_CHECKSUM}"
+                      BUILD_BYPRODUCTS "${BROTLI_STATIC_LIBRARY_ENC}"
+                                       "${BROTLI_STATIC_LIBRARY_DEC}"
+                                       "${BROTLI_STATIC_LIBRARY_COMMON}"
+                                       ${BROTLI_BUILD_BYPRODUCTS}
+                      CMAKE_ARGS ${BROTLI_CMAKE_ARGS}
+                      STEP_TARGETS headers_copy
+                      ${BROTLI_EP_OPTIONS})
+  if(CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
+    # Copy the libraries to our install directory manually.
     set(BROTLI_BUILD_DIR ${CMAKE_CURRENT_BINARY_DIR}/brotli_ep-prefix/src/brotli_ep-build)
     set(BROTLI_BUILD_LIBS
         "${BROTLI_BUILD_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}brotlienc-static${CMAKE_STATIC_LIBRARY_SUFFIX}"
@@ -1432,18 +1449,6 @@ macro(build_brotli)
 
     set(BROTLI_BUILD_INCLUDE_DIR
         ${CMAKE_CURRENT_BINARY_DIR}/brotli_ep-prefix/src/brotli_ep/c/include/brotli)
-
-    externalproject_add(brotli_ep
-                        ${EP_COMMON_OPTIONS}
-                        URL ${BROTLI_SOURCE_URL}
-                        URL_HASH "SHA256=${ARROW_BROTLI_BUILD_SHA256_CHECKSUM}"
-                        BUILD_BYPRODUCTS "${BROTLI_STATIC_LIBRARY_ENC}"
-                                         "${BROTLI_STATIC_LIBRARY_DEC}"
-                                         "${BROTLI_STATIC_LIBRARY_COMMON}"
-                                         ${BROTLI_BUILD_BYPRODUCTS}
-                        CMAKE_ARGS ${BROTLI_CMAKE_ARGS}
-                        STEP_TARGETS headers_copy
-                        INSTALL_COMMAND "")
     add_custom_command(TARGET brotli_ep
                        POST_BUILD
                        COMMAND ${CMAKE_COMMAND} -E copy_if_different
@@ -1451,17 +1456,6 @@ macro(build_brotli)
                                ${BROTLI_LIB_DIR}
                        COMMAND ${CMAKE_COMMAND} -E copy_directory
                                ${BROTLI_BUILD_INCLUDE_DIR} ${BROTLI_INCLUDE_DIR}/brotli)
-  else() # not emscripten - just behave as normal
-    externalproject_add(brotli_ep
-                        ${EP_COMMON_OPTIONS}
-                        URL ${BROTLI_SOURCE_URL}
-                        URL_HASH "SHA256=${ARROW_BROTLI_BUILD_SHA256_CHECKSUM}"
-                        BUILD_BYPRODUCTS "${BROTLI_STATIC_LIBRARY_ENC}"
-                                         "${BROTLI_STATIC_LIBRARY_DEC}"
-                                         "${BROTLI_STATIC_LIBRARY_COMMON}"
-                                         ${BROTLI_BUILD_BYPRODUCTS}
-                        CMAKE_ARGS ${BROTLI_CMAKE_ARGS}
-                        STEP_TARGETS headers_copy)
   endif()
 
   add_dependencies(toolchain brotli_ep)
```

kou commented on ci/docker/ubuntu-22.04-cpp.dockerfile:
```diff
@@ -114,10 +115,20 @@ RUN apt-get update -y -q && \
         tzdata \
-        wget && \
+        wget \
+        xz-utils  && \
```
```suggestion
        xz-utils && \
```

kou reviewed (commented):
> I've done the updates to enable tests correctly on emscripten - do you want them in this PR or in a separate one?

I suggest that we use a separated PR. Because they aren't small. If we we use a separated PR for them, we will be able to merge this faster.

joemarshall commented on ci/scripts/cpp_build.sh:
```diff
@@ -91,111 +91,139 @@ esac
-  ${ARROW_CMAKE_ARGS} \
-  ${source_dir}
+if [ "${ARROW_EMSCRIPTEN:-OFF}" = "ON" ]; then
+  if [ "${ARROW_BUILD_TYPE:-debug}" = "debug" ]; then
+    echo "Forcing non-parallel build for emscripten debug"
+    export CMAKE_BUILD_PARALLEL_LEVEL=1 
+    # emscripten debug linking takes *tons* of memory
+    # https://github.com/WebAssembly/binaryen/issues/4261
+    # so stop parallel builds for debug build
+    # or else crossbow CI runs out of memory
+  fi
+  
```
I think not, good spot. I'll test it.

jorisvandenbossche commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-emscripten

jorisvandenbossche commented:
@joemarshall there is a Python failure that seems to happen consistently for the different builds / different last commits:

```
__________________ TestThreadedCSVTableRead.test_cancellation __________________

self = <pyarrow.tests.test_csv.TestThreadedCSVTableRead object at 0x7f1cf2cbe760>

    def test_cancellation(self):
        if (threading.current_thread().ident !=
                threading.main_thread().ident):
            pytest.skip("test only works from main Python thread")
        # Skips test if not available
        raise_signal = util.get_raise_signal()
        signum = signal.SIGINT
    
        def signal_from_thread():
            # Give our workload a chance to start up
            time.sleep(0.2)
            raise_signal(signum)
    
        # We start with a small CSV reading workload and increase its size
        # until it's large enough to get an interruption during it, even in
        # release mode on fast machines.
        last_duration = 0.0
        workload_size = 100_000
        attempts = 0
    
        while last_duration < 5.0 and attempts < 10:
            print("workload size:", workload_size)
            large_csv = b"a,b,c\n" + b"1,2,3\n" * workload_size
            exc_info = None
    
            try:
                # We use a signal fd to reliably ensure that the signal
                # has been delivered to Python, regardless of how exactly
                # it was caught.
                with util.signal_wakeup_fd() as sigfd:
                    try:
                        t = threading.Thread(target=signal_from_thread)
                        t.start()
                        t1 = time.time()
                        try:
>                           self.read_bytes(large_csv)

opt/conda/envs/arrow/lib/python3.10/site-packages/pyarrow/tests/test_csv.py:1419: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
opt/conda/envs/arrow/lib/python3.10/site-packages/pyarrow/tests/test_csv.py:689: in read_bytes
    return self.read_csv(pa.py_buffer(b), **kwargs)
opt/conda/envs/arrow/lib/python3.10/site-packages/pyarrow/tests/test_csv.py:684: in read_csv
    table = read_csv(csv, *args, **kwargs)
pyarrow/_csv.pyx:1262: in pyarrow._csv.read_csv
    ???
pyarrow/_csv.pyx:1271: in pyarrow._csv.read_csv
    ???
pyarrow/error.pxi:154: in pyarrow.lib.pyarrow_internal_check_status
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   pyarrow.lib.ArrowCancelled: Operation cancelled. Detail: received signal 2

pyarrow/error.pxi:91: ArrowCancelled
```

joemarshall commented:
@jorisvandenbossche That is weird - I'm trying to reproduce it here. I can't see what has changed that isn't behind a `#ifdef EMSCRIPTEN`

kou commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -702,3 +714,47 @@ if(MSVC)
+    set(CMAKE_EXE_LINKER_FLAGS "${ARROW_EMSCRIPTEN_LINKER_FLAGS}  -sALLOW_MEMORY_GROWTH")
+  endif()
+
+  # limit debug info because building with DWARF debug info requires
+  # absolutely tons of memory
+  # https://github.com/WebAssembly/binaryen/issues/4261
+  if(${UPPERCASE_BUILD_TYPE} STREQUAL "DEBUG")
+    string(APPEND CMAKE_CXX_FLAGS_DEBUG " -g2")
+  endif()
```
Can we remove this because we've added https://github.com/apache/arrow/pull/37821/files#r1364914418 ?

kou reviewed (commented):
(I hope that I can re-review this in this week.)

kou commented on cpp/CMakePresets.json:
```diff
@@ -46,6 +46,37 @@
+        "ARROW_ACERO": "ON",
+        "ARROW_BUILD_SHARED": "OFF",
+        "ARROW_BUILD_STATIC": "ON",
+        "ARROW_BUILD_TESTS": "ON",
```
Could you remove this because we have it in `base-debug`?

```suggestion
```

kou commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -702,3 +714,47 @@ if(MSVC)
+  if(ARROW_TESTING)
+    # flags for building test executables for use in node
+    set(CMAKE_EXE_LINKER_FLAGS
+        "${ARROW_EMSCRIPTEN_LINKER_FLAGS}  -sERROR_ON_WASM_CHANGES_AFTER_LINK=1 -sALLOW_MEMORY_GROWTH -lnodefs.js -lnoderawfs.js --pre-js ${BUILD_SUPPORT_DIR}/emscripten-test-init.js"
```
Why do we need to use "linker" flags to set environment variables?
Test executables built with Emscripten can't read environment variables at runtime?

joemarshall replied:
the -s flags aren't environment flags, they're emscripten custom link flags. The `ERROR_ON_WASM_CHANGES_AFTER_LINK` makes sure that it is a debug build and isn't going through wasm_finalize, which slows everything down and isn't needed for test executables, and `ALLOW_MEMORY_GROWTH` tells the linker that there isn't a fixed memory size for the test.

joemarshall replied:
Ah. It's because the tests are run in node.js  When you build in emscripten, you can pass a pre-run javascript which is run on startup of the test executable to setup the environment the tests are run in. Without this, ARROW_TEST_DATA and PARQUET_TEST_DATA aren't passed from the containing environment into the emscripten webassembly module. I don't really know why it doesn't pass things through by default - I suspect it is something to do with not wanting to pass things that are machine specific through to the webassembly virtual machine, but this needs to be done so that tests can see those environment variables.

kou replied:
Ah, thanks. I understand.
I thought that we can set `ARROW_TEST_DATA`/`PARQUET_TEST_DATA` by passing something (environment variables or command line arguments) to test executables. But we can't do this without changing `*-test.js` by passing `--pre-js ...`.

kou commented on cpp/src/arrow/io/file_test.cc:
```diff
@@ -540,6 +544,9 @@ class TestPipeIO : public ::testing::Test {
 };
 
 TEST_F(TestPipeIO, TestWrite) {
+#ifdef __EMSCRIPTEN__
```
Which should we use? `EMSCRIPTEN`? `__EMSCRIPTEN__`?

joemarshall replied:
I checked the docs and they only say `__EMSCRIPTEN__` so I've made it always be that.

kou commented on cpp/src/arrow/util/rle_encoding_test.cc:
```diff
@@ -33,6 +33,8 @@
 #include "arrow/util/io_util.h"
 #include "arrow/util/rle_encoding.h"
 
+#include <stdio.h>
```
Could you revert needles changes such as this `include` and other empty line changes?

kou commented on cpp/src/arrow/util/rle_encoding_test.cc:
```diff
@@ -214,7 +216,12 @@ TEST(BitUtil, RoundTripIntValues) {
 void ValidateRle(const std::vector<int>& values, int bit_width,
                  uint8_t* expected_encoding, int expected_len) {
   const int len = 64 * 1024;
+#ifdef EMSCRIPTEN
+  // on Emscripten, this buffer won't fit in the stack
+  static uint8_t buffer[len];
+#else
   uint8_t buffer[len];
+#endif
```
How about using `std::vector buffer(len)` to allocate memory dynamically?

kou commented on cpp/src/arrow/util/rle_encoding_test.cc:
```diff
@@ -227,10 +234,9 @@ void ValidateRle(const std::vector<int>& values, int bit_width,
     EXPECT_EQ(encoded_len, expected_len);
   }
-  if (expected_encoding != NULL) {
+  if (expected_encoding != NULL && encoded_len == expected_len) {
```
Why do we need this?

joemarshall replied:
that's a bug in the test which I came across - because it uses `EXPECT_EQ`  as opposed to `ASSERT_EQ` above when checking the two lengths are the same , then whilst the test fails, the function continues to run - so in the case an incorrect length is returned, `memcmp` still runs and is comparing beyond the length of one of the buffers, and causing the whole test process to crash.

kou replied:
OK. Let's change the above `EXPECT_EQ()` to `ASSERT_EQ()` and remove the added condition.

jorisvandenbossche commented:
>> there is a Python failure that seems to happen consistently for the different builds / different last commits:
> 
> That is weird - I'm trying to reproduce it here. I can't see what has changed that isn't behind a `#ifdef EMSCRIPTEN`

I can reproduce it locally. When fetching this branch and building it in my existing development environment, I get the same failures.

I am not sure if it's something we can fix in the test, or requires an actual code fix. The current test assumes that it gets a `KeyboardInterrupt` exception. But it seems we now get a different error, a `pyarrow.lib.ArrowCancelled`, and this isn't caught in the test:

https://github.com/apache/arrow/blob/e62ec62e40b04b0bfce76d58369845d3aa96a419/python/pyarrow/tests/test_csv.py#L1418-L1422

If I change the test to also catch `ArrowCancelled` in addition to KeyboardInterrupt, it still does not pass, because the current code code to detect that the "correct" error was raised and breaks from the loop no longer does the trick:

https://github.com/apache/arrow/blob/e62ec62e40b04b0bfce76d58369845d3aa96a419/python/pyarrow/tests/test_csv.py#L1433-L1439

Because `exc_info.__context__` is now None, in contrast to what the test expects. 

I am not super familiar with the code that translates C++ exceptions into pyarrow/python exceptions, so not directly an idea why this PR affects that.

jorisvandenbossche commented:
It might also be that the test needs to be skipped when running on emscripten? The test is using stdlib `signal.raise_signal` (and the python docs mention that for emscripten "signals are emulated and therefore behave differently."), `threading` (this is also not available on emscripten?) and `socket` (also not available on emscripten according to the python docs).

jorisvandenbossche commented:
OK, looking into the only affected code that is related, i.e. `SendSignalToThread`, it seems the `#if` is not working. If I remove that locally, it works again. So for some reason, `#if !ARROW_ENABLE_THREADING` evaluates to True on a default build with threading enabled.

joemarshall commented:
Oh blooming heck - that should be `#ifndef ARROW_ENABLE_THREADING` not `#if !ARROW_ENABLE_THREADING`. I cannot understand why it doesn't break on main though.

jorisvandenbossche commented:
> that should be `#ifndef ARROW_ENABLE_THREADING` not `#if !ARROW_ENABLE_THREADING`.

I have been testing that locally, and that also doesn't seem to be working for me ..

jorisvandenbossche commented:
There is a `#include "arrow/util/config.h"` missing!

jorisvandenbossche commented on cpp/src/arrow/util/io_util.cc:
```diff
@@ -2058,7 +2063,9 @@ Status SendSignal(int signum) {
-#ifdef _WIN32
+#ifndef ARROW_ENABLE_THREADING
+  return Status::NotImplemented("Can't send signal with no threads");
+#elif _WIN32
```
```suggestion
#elif defined(_WIN32)
```

jorisvandenbossche commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-emscripten

jorisvandenbossche commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-emscripten

joemarshall commented:
@jorisvandenbossche it's running out of memory on crossbow - the test build output is 18GB, because it is statically linked and every test has all of arrow that it uses. 

It would be possible to build a shared wasm for testing only - i.e. run `emcc libarrow.a -o libarrow.wasm -sSIDE_MODULE=1 -g2` and then replace libarrow.a for libarrow.wasm in `ARROW_TEST_LINK_LIBS`? Bit of a hack

jorisvandenbossche commented:
cc @raulcd in case you know of other ways to increase the available memory (I know we regularly run into that issue on other builds as well?)

jorisvandenbossche commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-emscripten

kou reviewed (commented):
I could build (including test executables) on local but tests are still failed.
We can build but tests are failed on CI too.

How about moving test related changes to a separated PR and merging the CMake build part as the first step? (I think that we can merge the CMake build part with some additional simplications.)

joemarshall commented on cpp/cmake_modules/BuildUtils.cmake:
```diff
@@ -742,6 +742,14 @@ function(ADD_TEST_CASE REL_TEST_NAME)
                  --error-exitcode=1 ${TEST_PATH} ${ARG_TEST_ARGUMENTS}")
   elseif(WIN32)
     add_test(${TEST_NAME} ${TEST_PATH} ${ARG_TEST_ARGUMENTS})
+  elseif(CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
+    add_test(${TEST_NAME}
+             ${BUILD_SUPPORT_DIR}/run-test.sh
+             ${CMAKE_BINARY_DIR}
+             test
+             ${CMAKE_CROSSCOMPILING_EMULATOR}
+             ${TEST_PATH}
+             ${ARG_TEST_ARGUMENTS})
```
I see that, but don't we need the stuff in `run-test.sh`? I don't mind skipping it, but I don't see a reason not to use it for emscripten also?

joemarshall commented on cpp/cmake_modules/EmscriptenOverrides.cmake:
```diff
@@ -0,0 +1,25 @@
+set_property(GLOBAL PROPERTY TARGET_SUPPORTS_SHARED_LIBS TRUE)
+
+# stripping doesn't work on emscripten
+set(CMAKE_STRIP FALSE)
```
okay, doing it.

joemarshall commented:
@kou done those review changes, with the exception of the one I left open. I built it using `archery docker run ubuntu-cpp-emscripten` and it builds and tests okay on my machine now. I had to turn down some concurrency options in both build and test, or else it did weird things in docker, but I think it is worth firing off a crossbow build now to check whether it tests okay (@jorisvandenbossche)

jorisvandenbossche commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-emscripten

jorisvandenbossche commented:
The crossbow is working now. 

@kou do you have further comments here? (apart from updating with latest main to fix the conflicts)

jorisvandenbossche commented:
(once updated, we should also do a run of all nightly builds)

kou commented:
Sorry. I'll review this in a few days.

kou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1349,6 +1366,13 @@ macro(build_snappy)
+    # ignore linker flag errors, as snappy sets
+    # -Werror -Wall, and emscripten doesn't support -soname
+    list(APPEND SNAPPY_CMAKE_ARGS
+         "-DCMAKE_SHARED_LINKER_FLAGS=${CMAKE_SHARED_LINKER_FLAGS} -Wno-error=linkflags")
```
```suggestion
         "-DCMAKE_SHARED_LINKER_FLAGS=${CMAKE_SHARED_LINKER_FLAGS}" "-Wno-error=linkflags")
```

kou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1349,6 +1366,13 @@ macro(build_snappy)
   endif()
 
+  if(CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
+    # ignore linker flag errors, as snappy sets
```
```suggestion
    # ignore linker flag errors, as Snappy sets
```

kou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1349,6 +1366,13 @@ macro(build_snappy)
 
+  if(CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
+    # ignore linker flag errors, as snappy sets
+    # -Werror -Wall, and emscripten doesn't support -soname
```
```suggestion
    # -Werror -Wall, and Emscripten doesn't support -soname
```

kou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1798,6 +1859,38 @@ macro(build_protobuf)
+                          PROPERTIES IMPORTED_LOCATION "${PROTOBUF_HOST_COMPILER}")
+
+    add_dependencies(arrow::protobuf::host_protoc protobuf_ep_host)
+
```
```suggestion
```

kou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1798,6 +1859,38 @@ macro(build_protobuf)
+    add_dependencies(arrow::protobuf::host_protoc protobuf_ep_host)
+
+  endif()
+
```
```suggestion
```

kou commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -702,3 +730,36 @@ if(MSVC)
+  if(ARROW_TESTING)
+    # flags for building test executables for use in node
+    set(CMAKE_EXE_LINKER_FLAGS
+        "${ARROW_EMSCRIPTEN_LINKER_FLAGS}  -sERROR_ON_WASM_CHANGES_AFTER_LINK=1 -sALLOW_MEMORY_GROWTH -lnodefs.js -lnoderawfs.js --pre-js ${BUILD_SUPPORT_DIR}/emscripten-test-init.js"
```
```suggestion
        "${ARROW_EMSCRIPTEN_LINKER_FLAGS} -sERROR_ON_WASM_CHANGES_AFTER_LINK=1 -sALLOW_MEMORY_GROWTH -lnodefs.js -lnoderawfs.js --pre-js ${BUILD_SUPPORT_DIR}/emscripten-test-init.js"
```

kou commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -702,3 +730,36 @@ if(MSVC)
+  else()
+    set(CMAKE_EXE_LINKER_FLAGS "${ARROW_EMSCRIPTEN_LINKER_FLAGS}  -sALLOW_MEMORY_GROWTH")
+  endif()
+
```
```suggestion
```

kou commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -702,3 +730,36 @@ if(MSVC)
+        "${ARROW_EMSCRIPTEN_LINKER_FLAGS}  -sERROR_ON_WASM_CHANGES_AFTER_LINK=1 -sALLOW_MEMORY_GROWTH -lnodefs.js -lnoderawfs.js --pre-js ${BUILD_SUPPORT_DIR}/emscripten-test-init.js"
+    )
+  else()
+    set(CMAKE_EXE_LINKER_FLAGS "${ARROW_EMSCRIPTEN_LINKER_FLAGS}  -sALLOW_MEMORY_GROWTH")
```
```suggestion
    set(CMAKE_EXE_LINKER_FLAGS "${ARROW_EMSCRIPTEN_LINKER_FLAGS} -sALLOW_MEMORY_GROWTH")
```

kou commented on cpp/src/arrow/util/mutex.cc:
```diff
@@ -35,9 +36,12 @@ struct Mutex::Impl {
 
 Mutex::Guard::Guard(Mutex* locked)
     : locked_(locked, [](Mutex* locked) {
+#if !defined(__EMSCRIPTEN__) || defined(ARROW_ENABLE_THREADING)
```
Is this `__EMSCRIPTEN__` check redundant?

```suggestion
#ifdef ARROW_ENABLE_THREADING
```

kou commented on cpp/src/arrow/util/rle_encoding_test.cc:
```diff
@@ -214,7 +214,14 @@ TEST(BitUtil, RoundTripIntValues) {
+#ifdef __EMSCRIPTEN__
+  // don't make this on the stack as it is
+  // too big for emscripten
+  std::vector<uint8_t> buffer_vec((size_t)len);
```
```suggestion
  std::vector<uint8_t> buffer_vec(static_cast<size_t>(len));
```

kou commented on cpp/src/arrow/util/rle_encoding_test.cc:
```diff
@@ -256,7 +263,14 @@ void ValidateRle(const std::vector<int>& values, int bit_width,
+#ifdef __EMSCRIPTEN__
+  // don't make this on the stack as it is
+  // too big for emscripten
+  std::vector<uint8_t> buffer_vec((size_t)len);
```
```suggestion
  std::vector<uint8_t> buffer_vec(static_cast<size_t>(len));
```

kou commented on docs/source/developers/cpp/emscripten.rst:
```diff
@@ -0,0 +1,100 @@
+.. _developers-cpp-emscripten:
+#################################################
+Cross compiling for WebAssembly with Emscripten
+#################################################
```
It seems that we use different marker character (e.g. `#` for top level) for other documents such as https://raw.githubusercontent.com/apache/arrow/main/docs/source/developers/cpp/building.rst (e.g. `=` for top level) in the same directory.
Could you use the same style?

jorisvandenbossche commented:
@joemarshall do you have time to update for the last comments from @kou?

joemarshall commented:
I'll do my best. It's teaching term here and I've got a bunch of projects running, but it doesn't look like massive changes.

joemarshall commented:
@jorisvandenbossche I did those comments. Some go tests seem to be sporadically failing, i.e. they didn't fail before I did cmake_format, then another commit and they fail. I don't think it is me, but if you could hit rerun on them to check. 

Other than that no comments left unaddressed now.

joemarshall commented:
@jorisvandenbossche @kou Weird - github lost a commit somewhere - I had to make a couple of test commits to make it realise that I'd fixed the lint errors. Anyway, should be there now... There's a fail in R which I think isn't me.

joemarshall commented:
@kou I just merged in the latest changes and the CI seems to have mostly got back to normal now. If there's any chance we could merge now that would be super.

kou commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-emscripten

kou commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -692,18 +699,39 @@ if(NOT MSVC)
+        string(APPEND C_RELWITHDEBINFO_FLAGS " -ggdb")
+        string(APPEND CXX_RELWITHDEBINFO_FLAGS " -ggdb")
+      endif()
+
```
```suggestion
```

kou commented on cpp/cmake_modules/SetupCxxFlags.cmake:
```diff
@@ -692,18 +699,39 @@ if(NOT MSVC)
+      endif()
+
     endif()
+
```
```suggestion
```

kou commented on cpp/src/arrow/compute/kernels/scalar_string_test.cc:
```diff
@@ -21,6 +21,7 @@
 #include <vector>
 
 #include <gmock/gmock.h>
+#include <gtest/gtest-spi.h>
```
```suggestion
```

kou commented on cpp/src/arrow/io/file_test.cc:
Could you add `#include "arrow/util/future.h"` for `ARROW_ENABLE_THREADING`?

kou commented on cpp/src/arrow/util/atfork_test.cc:
Could you add `#include "arrow/util/future.h"` for `ARROW_ENABLE_THREADING`?

kou commented on cpp/src/arrow/util/cache_test.cc:
Could you add `#include "arrow/util/future.h"` for `ARROW_ENABLE_THREADING`?

kou commented on cpp/src/arrow/util/cancel_test.cc:
Could you add `#include "arrow/util/future.h"` for `ARROW_ENABLE_THREADING`?

kou commented on cpp/src/arrow/util/counting_semaphore_test.cc:
Could you add `#include "arrow/util/future.h"` for `ARROW_ENABLE_THREADING`?

kou commented on cpp/src/arrow/util/future_test.cc:
Could you add `#include "arrow/util/future.h"` for `ARROW_ENABLE_THREADING`?

kou commented on cpp/src/arrow/util/io_util_test.cc:
Could you add `#include "arrow/util/future.h"` for `ARROW_ENABLE_THREADING`?

joemarshall commented:
@kou fixed the emscripten test failures and done those comments now.

kou commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-emscripten

jorisvandenbossche commented:
The failing integration test build seems unrelated (it's failing in a rust<->rust test)

bitsondatadev commented:
> The failing integration test build seems unrelated (it's failing in a rust<->rust test)

@jorisvandenbossche and @kou, are there any changes you're waiting for or is this ready for another review?

kou commented:
Sorry... I should have said `#include "arrow/util/config.h"` not `#include "arrow/util/future.h"` for https://github.com/apache/arrow/pull/37821#discussion_r1542212438 and so on...

Could you use `arrow/util/config.h` not `arrow/util/future.h` instead...? (Can I push a commit to this branch?)

joemarshall commented:
> Sorry... I should have said `#include "arrow/util/config.h"` not `#include "arrow/util/future.h"` for https://github.com/apache/arrow/pull/37821#discussion_r1542212438 and so on...
> 
> Could you use `arrow/util/config.h` not `arrow/util/future.h` instead...? (Can I push a commit to this branch?)

Yes pushing a commit is fine.

kou commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-emscripten

kou commented:
@github-actions crossbow submit -g cpp

kou reviewed (approved):
+1

bitsondatadev commented:
Thanks for your reviews @kou  and thanks @joemarshall for all the hard work!!

joemarshall commented:
Brilliant, thanks everyone

raulcd commented:
@github-actions crossbow submit test-build-vcpkg-win

raulcd commented:
I am trying to understand if this PR was the one causing the `test-build-vcpkg-win` job failure.

pitrou commented on ci/docker/ubuntu-22.04-cpp.dockerfile:
```diff
@@ -115,10 +116,20 @@ RUN apt-get update -y -q && \
 
+# install emscripten using EMSDK
+ARG emscripten_version="3.1.45"
+RUN cd ~ && git clone https://github.com/emscripten-core/emsdk.git && \
```
Hmm, two things:
1. it would have been better for resource usage to make a separate dockerfile for this, perhaps
2. could this be a script to avoid having tons of hand-coded bash code in dockerfiles?

pitrou commented on ci/docker/ubuntu-22.04-cpp.dockerfile:
```diff
@@ -151,6 +162,9 @@ RUN if [ "${gcc_version}" = "" ]; then \
     fi
 
+# make sure zlib is cached in the EMSDK folder
+RUN source ~/emsdk/emsdk_env.sh && embuilder --pic build zlib
```
Any reason why this isn't together with the install step above?
